### PR TITLE
Add `ClientRequest` and `http.request()`

### PR DESCRIPTION
**Pull request by yassernasc**
Main changes:

- At HTTPOutgoingMessage's `flushHeader` method, the logic to construct the header was moved to the child classes (ServerResponse and ClientRequest) by the method `_arrangeHeader`.

- At HTTPClientRequest's `_ondata` method, we have an implementation of a HTTP response parsing, which includes catching the Message Body; that message is pushed to the HTTPIncomingMessage, I think it's missing a 'stream drain' handling there.

- Still about the client `_ondata` method, by the end of it I am destroying the socket manually, IDK if it's the right approach. I've added it to get things working for the `example/request.js` code sample.

---

Creating the PR as `draft` to get early feedback. I am validating more use cases and the unit tests are going to be updated.

### CODE REVIEW

File: lib/client-request.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    const blankLine = Buffer.from([13, 10, 13, 10])
+    const i = data.indexOf(blankLine)
+
+    if (i === -1) return this._socket.destroy()
+
+    const head = data.subarray(0, i)
+    const messageBody = data.subarray(i + blankLine.byteLength)
+
+    this._onresponse(head, messageBody)
+  }
```
Comment (kasperisager) [inline lib/client-request.js:69]: This assumes that a single TCP message corresponds to a full HTTP header, which isn't necessarily the case. TCP doesn't frame messages and is free to split writes into as many or few messages as it finds appropriate.

You can see an example of how to correctly stitch together the TCP messages in the `ServerConnection` class, whose `_ondata()` handler uses a persistent buffer (`_buffer`) to gradually build up the complete HTTP header and only then forwards to the  `_onrequest()` handler.

File: lib/client-request.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+          }
+        }
+
+        res.push(message)
+      } else {
+        res.push(body)
+      }
+    }
+
+    res.push(null)
```
Comment (kasperisager) [inline lib/client-request.js:120]: Similarly, this assumes that the full HTTP body is available immediately, which isn't necessarily the case. It may arrive incrementally across several TCP messages _after_ the HTTP header has arrived.

**Review by kasperisager — CHANGES_REQUESTED** [review]

**Comment by yassernasc** [comment]
I managed to implement a solution to receive the information on multiple pieces but at the end of the day, my solution proved to be painfully slow. I could make the test `server does a big write` work but decreasing a lot the message size. I am pretty sure that the slow part is the line `message = Buffer.concat([message, Buffer.from([b])])` inside the for loop (if I comment that line, the test doesn't get completely stuck), which is executed for each byte from the accumulated server response. Maybe I am doing bad buffer manipulations or more likely, the algorithm itself is impracticable (doesn't scale).

File: lib/client-request.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      let message = Buffer.from([])
+
+      let readingMessage = false
+      let hits = 0
+
+      for (let i = 0; i < this._responseBodyBuffer.byteLength; i++) {
+        const b = this._responseBodyBuffer[i]
+
+        if (readingMessage && hits === 0 && b !== 13) {
+          message = Buffer.concat([message, Buffer.from([b])])
```
Comment (kasperisager) [inline lib/client-request.js:129]: This is indeed the bottleneck as it builds up the buffer in `O(n^2)` time. Instead of incrementally appending bytes to the message buffer, you can keep track of where the current chunk started and then grab the entire chunk using `subarray(start, i)` as soon as the chunk ends.

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: lib/client-request.js (hunk 3 above)
Comment (kasperisager) [reply lib/client-request.js:129]: This is essentially what https://github.com/holepunchto/bare-http1/blob/902e63cf02ad24341e5c5156a66e5bfa27bb5279/lib/server-connection.js#L53-L55 does as well, but it also splits the remainder so the current chunk always starts at 0. That's less state to keep track of.

Comment (yassernasc) [reply lib/client-request.js]: great, I am using `Buffer.concat` just once and it has improved performance a lot.

File: lib/client-request.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+      const messageSlices = slices.map(([s, e]) => this._responseBodyBuffer.subarray(s, e))
+      this._responseBodyBuffer = Buffer.concat(messageSlices)
+    }
+
+    if ((!chunked && this._responseBodyBuffer.byteLength === len) || hasChunkedTail) {
+      if (this._responseBodyBuffer.byteLength > 0) this._response.push(this._responseBodyBuffer)
+      this._response.push(null)
+      this._socket.destroy()
+    }
```
Comment (kasperisager) [inline lib/client-request.js:145]: This will read the body in its entirety before pushing it to the stream, which isn't ideal as it will buffer the entire body in memory before passing it along. Instead, we should push chunks as they arrive, buffering only the bare minimum between TCP fragments.

**Review by kasperisager — CHANGES_REQUESTED** [review]
It's progressing! 👏

File: lib/client-request.js (hunk 4 above)
Comment (yassernasc) [reply lib/client-request.js]: That's true, I pushed a new commit and the current behavior is:
- For requests with `content-length`: Any data available is sent immediately.
- For chunked messages: Chunks are sent when they are complete, individually, that is, big chunks will be kept in memory similarly as before, the unit test for receiving big writes has only 3 chunks, for instance.

I am trying a new implementation to send "chunks in chunks" but the state management + parsing lines + edge cases handling is getting a bit wild.

**Review by kasperisager — CHANGES_REQUESTED** [review]
It's looking really solid now! 🙌 I refactored your parsing code in the same style as `HTTPServerConnection` and likewise added a `HTTPClientConnection` class which should also pave the way for `Agent` support down the road.

There's now a pretty significant overlap between `HTTPServerConnection` and `HTTPClientConnection` as both perform HTTP parsing. This should eventually converge in a proper HTTP parsing abstraction, ideally implemented in native code at some point.

I think the only missing bit now is support for `http.request(url[, options][, callback])`: https://nodejs.org/api/http.html#httprequesturl-options-callback. Bare has a global `URL.isURL()` method you can use for that.

File: lib/constants.js [hunk 5]
```diff
@@ -1,4 +1,11 @@
 module.exports = {
+  state: {
+    BEFORE_HEAD: 1,
+    IN_HEAD: 2,
+    IN_BODY: 3,
+    BEFORE_CHUNK: 4,
+    IN_CHUNK: 5
+  },
```
Comment (kasperisager) [inline lib/constants.js:8]: This is basically a condensed version of the previous `readingEnum` and the various `HTTPClientRequest` flags.

Comment (yassernasc) [reply lib/constants.js:8]: good insights there, thanks

Comment (yassernasc) [reply lib/constants.js:8]: Support for `http.request(url[, options][, callback])` added. Also, new unit tests to cover that.

**Review by kasperisager — APPROVED** [review]
I simplified the argument juggling a bit and move the overload to `http.request()` to keep the `ClientRequest` constructor simple. In general, argument overloading is easier to manage if one first shuffles the callback argument into place and then start shuffling the remaining arguments.

I also added default port handling so `https:` URLs, for example, use port 443 unless otherwise specified.

Let's land it! 🚀

**Comment by yassernasc** [comment]
lol

The argument refactor was damn good, a lot simpler; just a small thing, maybe there is a incompatibility with node.js, following the docs: `If both 'url' and 'options' are specified, the objects are merged, with the 'options' properties taking precedence.` So, I think that our precedence order is inverted.
